### ISSUE TITLE
Variables used in ex_committor.ipynb might be wrong

### ISSUE THREAD

**Issue by Alexander-the-Evil** [luigibonati/mlcolvar#174]
Hi, I was looking through the tutorial for ex_committor.ipynb. I am confused by your regex you use to filter out the training set variables.
As I understood initially, you want to use a set of 45 pairwise distances. However, if we look into the code you provide, you instead select scaled positions of 10 atoms (30 positions in total). The code then initializes a model with the input vector size of 45. What is the intended way to do this tutorial then, with distances (as I'd assumed) or with positions?
Best,
Alex

**Comment by Alexander-the-Evil** [comment]
Ah, ok, sorry, I am stupid. The conversion is done internally through PairwiseDistances. It would however require a constant cell size.
If NPT ensemble is used, apparently we cannot do that anymore. Am I correct to assume that, in NPT, there is no way to use this pipeline (positions -> PairwiseDistances -> smart datasets), and instead I should calculate pairwise distances from the trajectory and feed them to smart datasets directly?

Best,
Alex

**Comment by EnricoTrizio** [comment]
Ciao @Alexander-the-Evil, in the committor model, the variational part of the loss function depends on the square modulus of the gradients of the committor **wrt the atomic positions.**
Our NN takes as inputs some physical descriptors (e.g., those 45 distances), but these need to be computed in Python from the positions of the involved atoms to have the gradients up to the positions.

So what we do is: positions (30) --> descriptors (45) --> nn (45 -> 1) --> committor (1)

As you can imagine, only the derivatives of the second part (i.e., descriptors --> nn --> committor) change during the optimization, whereas the ones of positions --> descriptors part are always the same. For this reason, we compute the latter only once and store them in an efficient way with that `SmartDerivatives` object, as they are very sparse.

Hope it's clearer now,

**Comment by EnricoTrizio** [comment]
> Am I correct to assume that, in NPT, there is no way to use this pipeline (positions -> PairwiseDistances -> smart datasets), and instead I should calculate pairwise distances from the trajectory and feed them to smart datasets directly?

You still need the dependency on the positions (see previous comment) and, for the time being, no, there is no support for NPT simulations or changing cells.
The transform modules are meant to use constant cells for now.

It's something that I'll do sooner or later. If you want to try, you're welcome to.

In the meantime, a workaround could be to center beforehand the part of the system you are interested in so that PBC/cells are not an issue anymore (hopefully).

Hope it helps.

**Comment by Alexander-the-Evil** [comment]
Thank you @EnricoTrizio !
Yes, that helps. I now have everything to run for the application scenarios I have in mind. I am very excited to do that soon!

For broader case space though, one last question. Would the training/usage in principle work without the "smart dataset" part? 

Thanks,
Alex

**Comment by EnricoTrizio** [comment]
You can skip the smart dataset thing and add the computation of the descriptors as a preprocessing module of your model (i.e., model.preprocessing = your_descriptor_calculation).

It works, but it is much slower and more memory-consuming.
